### DISCUSSION TITLE
How to use Lit customElement in an application without the element beeing part of the applications code-base

### DISCUSSION THREAD

**Discussion by zwif** [lit/lit#2298] [Q&A]
How to use a lit customElement in an Application without making it part of the application's code-base?

Consider following scenario:
There is an existing web-application (which is bundled/optimized via webpack).   
Additionally, we have a custom element developed from the [lit-elements-starter-ts](https://github.com/lit/lit-element-starter-ts) which is - [as recommended](https://github.com/lit/lit-element-starter-ts#:~:text=We%20recommend%20publishing%20components%20as%20unoptimized%20JavaScript%20modules%2C%20and%20performing%20build-time%20optimizations%20at%20the%20application%20level.) - published as unoptimized/unbundled js module.

Development of the application and the element is independent (they live in different git repos). Thus, to include the element to the application, we install it as npm dependency: `npm install <my-lit-element> --save`.  Somewhere in the application's template-code, it's used: `<my-lit-element></my-lit-element>`.

The naive approach to make the element work is to extend the existing webpack workflow of the application to copy the lit element's published code from the node-module to some place where the running application can access it. But since the lit element is neither optimized nor bundled in any way, this still needs to be done, obviously. This is the part where we are stuck and need advice!

We are reading the lit docs on [Building for Production](https://lit.dev/docs/tools/production/) and [Adding Lit to an Existing Application](https://lit.dev/docs/tools/adding-lit/), but can't find the information required to get this working. Keep in mind:
  1. Don't want to use rollup since the application already uses webpack.
  2. Don't want to include the lit element source code in the application source code.

In my opinion, there are two possible approaches to get this working:
 1. Disrespect the Recommendations and publish the lit element in a production-ready way -> minimized with its dependencies bundled. _Probably not a good idea, e.g. when multiple custom elements are used and each is bundled with its dependencies we probably end up with some dependencies included in multiple bundles._
 2. Extend the Webpack Workflow of the application to bundle/optimize the lit element which lives in the node_modules. But how to do this? What could a webpack config look like?
 

Any help is appreciated!

**Comment by thescientist13** [comment]
Just to confirm, are you not pulling in any third party dependencies from npm at the moment with webpack?  Or are you experiencing an issue with bundling a custom element with webpack specifically?

Just curious I suppose if you could elaborate more on this, as I'm not quite sure I understand what you mean
> Don't want to include the lit element source code in the application source code.

Do you mean npm dependencies bundled with your app code, or that you mean copy / pasting the code into your repo?

FWIW, webpack has many bundling / chunking / optimization settings if the defaults are not what you need, but likely webpack will try and find the most common ancestor of a dependency and bundle / split things appropriately.  However you can also get more granular if you specifically want to "silo" a specific dependency / module into its own chunk / bundle at build time, if that's the output you are looking for.  But bundling npm dependencies is something it should do out of the box since it knows how to resolve _node_modules_ intuitively.

**Comment by justinfagnani** [comment]
By installing the element via npm and importing it into your application, WebPack should already be bundling the element. Are you excluding `node_modules/` from compilation in WebPack?

**Comment by zwif** [comment]
Thanks for the quick replies! 

@thescientist13 
> Do you mean npm dependencies bundled with your app code, or that you mean copy / pasting the code into your repo?

Actually both - well, more or less at least. I do not want the code in my repo. In order to have it work, I obviously need to have it bundled with my app code somehow - But I want a separate bundle for each element.

@justinfagnani 
> By installing the element via npm **and importing it** into your application, [...]

That's the thing - I am not importing it (thus webpack will not just know about it)  - the only reference from the application to the element is something like following in the application's template-code:
```
<script type="module" src="/lib/components/my-elements.js">
<my-element></my-element>
```
<br>
I see, this evolves more to a webpack-specific question as lit - but I was hoping to find some hints on how to do this in the lit docs...  

It looks like I have to adjust my webpack-config for each and every element in order to tell webpack to use an extra entry/chunk respectively - or do I miss something?

<hr>
But in general - is this such a rare use case (I could rarely find information about it online - leading me to ask this question)? I would expect it to be fairly common to just include an element as npm dependency and reference it only from the template of the application?! Or is just anyone so familiar with his bundler that the bundler-configuration is a no-brainer?

**Reply by thescientist13** [reply]
Most module bundlers like webpack will only bundle JS by default / out of the box, as fed to it in the form of an "entry" file like _src/index.js_, at which point it will then follow all `import` / `require` etc from there.

So yes, I guess I would say its another expectation entirely to think that it would also know how to crawl your HTML and do the same from a `<script>` tag.  However, some tools like [Parcel](https://parceljs.org/) do work this way.  They can use an _index.html_ _as_ the entry point and so then it _will_ crawl through your `<script>`, `<link>`, etc and bundle based on that.

Linking to _node_modules_ is another expectation altogether as well, as browsers have no concept of a _node_modules_ and the ESM spec actually expects your `import` statements to start with `.` to indicate a relative file path.  Auto magical linking for anything to _node_modules_ is going to be very tooling dependent since none of this is web standard, though it is getting there with things like [import maps](https://github.com/WICG/import-maps).

TL;DR you can probably do it with webpack since it is very capable and configurable and has a wide ecosystem of support, but since by default it only knows JS, you will need to find / write a plugin if you want it crawl your HTML and be able to link and bundle paths you define there.  But at this point, maybe it would easier to just `import` it in your app code?  If you are code splitting, then webpack should optimize for you to help reduce duplicate bundles.

----

FWIW, the [Lit docs do try to surface and expose](https://lit.dev/docs/tools/requirements/) those "magic tricks" that bundlers and frontend tools have buried under the surface, so it's totally understandable why a lot of this feels like such a black box or new.  It may be worth evaluating other tools that were born out of the era of ESM such as the [**Modern Web**](https://modern-web.dev/) suite, [**Vite**](https://vitejs.dev/), and [**Greenwood**](https://www.greenwoodjs.io/) (personal shameless plug) that might be more amenable to a web / HTML centric workflow.

Hope that helps, and to get into the nuance a bit more, might recommend joining the [Lit and WC adjacent community on slack](https://lit-and-friends.slack.com/join/shared_invite/zt-llwznvsy-LZwT13R66gOgnrg12PUGqw#/shared-invite/email) as well depending on where your journey takes you.  ✌️

**Reply by zwif** [reply]
Thanks for the information! With that given, I had the confidence that I have to dig more into the webpack configuration and got it working.

I will leave my solution for the record, but I'm pretty sure there are alternative solutions as well:

1. I defined an array of the names of all custom elements in my application.
2. I create a bundle for each custom element by [defining a cache-group in the webpack config](https://webpack.js.org/plugins/split-chunks-plugin/#splitchunkscachegroups) which matches the custom elements  (via `test`-keyword) : `test: /[\\/]node_modules[\\/](my-element|another-element)[\\/]/` (inspired from [this article](https://medium.com/hackernoon/the-100-correct-way-to-split-your-chunks-with-webpack-f8a9df5b7758)) - the `test`-regex is dynamically created from the names of point 1.
3. Now you can add an import in one of the webpack `entry`-files  (`import 'my-element`) and webpack injects the bundle as script in your template (you can just use the custom-tag in your template without caring about the `<script></script>` import of that element). 
4. (_Bonus_) Instead of just the hard import `import 'my-element` - I call a `bootstrapWebComponents`-function within one of the `entry`-files which searches the DOM for these custom-tags - since the names/tags are known (as of point 1.) - and [dynamically creates the import](https://webpack.js.org/api/module-methods/#dynamic-expressions-in-import) - resulting in the custom-elements bundles being lazy-loaded.

**Reply by thescientist13** [reply]
Nice, and ➕ 1 on having it lazy loaded too!  Thanks for leaving your steps behind for others to find.  🙌
